rust/psibase/src/main.rs

```diff
@@ -1014,6 +1070,115 @@ async fn upload_tree(args: &UploadArgs) -> Result<(), anyhow::Error> {
     Ok(())
 }
 
+async fn publish(args: &PublishArgs) -> Result<(), anyhow::Error> {
+    let (mut client, _proxy) = build_client(&args.node_args.proxy).await?;
+
+    // Read package metadata and determine which packages are already published
+    let mut query_existing = "query {".to_string();
+
+    let mut all_meta = Vec::new();
+
+    for (i, path) in args.packages.iter().enumerate() {
+        let f = File::open(path).with_context(|| format!("Cannot open {}", path.display()))?;
+        let package = PackagedService::new(BufReader::new(f))?;
+        let meta = package.meta();
+        use std::fmt::Write;
+        write!(
+            &mut query_existing,
+            " p{}: package(owner: {}, name: {}, version: {}) {{ sha256 }}",
+            i,
+            serde_json::to_string(&args.sender)?,
+            serde_json::to_string(&meta.name)?,
+            serde_json::to_string(&meta.version)?
+        )?;
+        all_meta.push(meta.clone());
+    }
+
+    query_existing += " }";
+
+    let existing = psibase::gql_query::<serde_json::Value>(
+        &args.node_args.api,
+        &mut client,
+        packages::SERVICE,
+        query_existing,
+    )
+    .await?;
+    let existing = existing.as_object().unwrap();
+    let num_new = existing.values().filter(|p| p.is_null()).count();
+
+    let progress = ProgressBar::new(num_new as u64).with_style(ProgressStyle::with_template(
+        "{wide_bar} {pos}/{len}\n{msg}",
+    )?);
+    progress.set_message("Preparing transactions");
+
+    let tapos = get_tapos_for_head(&args.node_args.api, client.clone()).await?;
+
+    let action_limit: usize = 1024 * 1024;
+    let mut builder = TransactionBuilder::new(
+        action_limit,
+        |actions: Vec<Action>| -> Result<SignedTransaction, anyhow::Error> {
+            Ok(sign_transaction(
+                with_tapos(&tapos, actions, &args.sig_args.proposer, false),
+                &args.sig_args.sign,
+            )?)
+        },
+    );
+
+    for (i, path) in args.packages.iter().enumerate() {
+        let mut f = File::open(path).with_context(|| format!("Cannot open {}", path.display()))?;
+        // Read the file
+        let mut content = Vec::new();
+        f.read_to_end(&mut content)?;
+        // Calculate the package hash
+        let hash: [u8; 32] = Sha256::digest(&content).into();
+        let existing_info = &existing[&format!("p{}", i)];
+        let meta = &all_meta[i];
+        if existing_info.is_null() {
+            let dest = format!("/packages/{}-{}.psi", &meta.name, &meta.version);
```

Comment:
Should `/packages/...` be noted somewhere as a reserved / well-known upload location? Users could accidentally have content get overshadowed if they expected to use a `/packages` path